File: Latihan/LatihanViewModel/java/app/src/main/java/com/dicoding/picodiploma/myviewmodel/MainActivity.java
package com.dicoding.picodiploma.myviewmodel;

import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

public class MainActivity extends AppCompatActivity {

    // Membuat varible global
    private EditText edtWidth, edtHeight, edtLength;
    private TextView tvResult;
    private MainViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Menghubungkan variable editText dan textView dengan layout
        edtWidth = findViewById(R.id.edt_width);
        edtHeight = findViewById(R.id.edt_height);
        edtLength = findViewById(R.id.edt_length);
        tvResult = findViewById(R.id.tv_result);

        // Menghubungkan ViewModel dengan Activity
        viewModel = new ViewModelProvider(this).get(MainViewModel.class);

        // Menampilkan hasil pertama kali
        displayResult();

        // Memberikan aksi klik kepada button calculate
        findViewById(R.id.btn_calculate).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {

                // Mendapatkan string dari editText
                String width = edtWidth.getText().toString();
                String height = edtHeight.getText().toString();
                String length = edtLength.getText().toString();

                // Melakukan pengecekan apakah empty atau tidak
                if (width.isEmpty()) {
                    edtWidth.setError("Masih kosong");
                } else if (height.isEmpty()) {
                    edtHeight.setError("Masih kosong");
                } else if (length.isEmpty()) {
                    edtLength.setError("Masih kosong");
                } else {

                    // Melakukan pengiriman string ke ViewModel
                    viewModel.calculate(width, height, length);

                    //Menampilkan hasil dari calculate
                    displayResult();
                }
            }
        });
    }

    // Menampilkan hasil dari ViewModel ke tvResults
    private void displayResult() {
        tvResult.setText(String.valueOf(viewModel.result));
    }
}

File: Latihan/LatihanViewModelTest/kotlin/app/src/main/java/com/dicoding/picodiploma/myviewmodel/MainActivity.kt
package com.dicoding.picodiploma.myviewmodel

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    // Membuat varibel global
    private lateinit var viewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        // Menghubungkan ViewModel dengan Activity
        viewModel = ViewModelProvider(this)[MainViewModel::class.java]
        // Menampilkan hasil pertama kali
        displayResult()
        // Memberikan aksi klik kepada button calculate
        btn_calculate.setOnClickListener {
            // Mendapatkan string dari editText
            val width = edt_width.text.toString()
            val height = edt_height.text.toString()
            val length = edt_length.text.toString()
            // Melakukan pengecekan apakah empty atau tidak
            when {
                width.isEmpty() -> {
                    edt_width.error = "Masih kosong"
                }
                height.isEmpty() -> {
                    edt_height.error = "Masih kosong"
                }
                length.isEmpty() -> {
                    edt_length.error = "Masih kosong"
                }
                else -> {
                    // Melakukan pengiriman string ke ViewModel
                    viewModel.calculate(width, height, length)
                    //Menampilkan hasil dari calculate
                    displayResult()
                }
            }
        }
    }

    // Menampilkan hasil dari ViewModel ke tvResults
    private fun displayResult() {
        tv_result.text = viewModel.result.toString()
    }
}